metron-interface/metron-rest/src/main/java/org/apache/metron/rest/controller/RestExceptionHandler.java

```diff
@@ -45,4 +45,14 @@ private HttpStatus getStatus(HttpServletRequest request) {
     }
     return HttpStatus.valueOf(statusCode);
   }
+
+  private String getFullMessage(Throwable ex) {
+    String fullMessage = ex.getMessage();
+    Throwable cause = ex.getCause();
+    while(cause != null) {
+      fullMessage = cause.getMessage();
+      cause = cause.getCause();
+    }
+    return fullMessage;
```

Comment:
Could we also get the stacktrace?  Maybe a call to `ExceptionUtils.getStackTrace(ex)` appended?

Reply:
Maybe be a security risk to show stack in ui or rest errors?  Could we control it by profile or something?

Reply:
Perhaps a good alternative would be to show the error to the endpoint (the UI) and at least log the exception in the metron-rest.log

Reply:
The stack trace is not shown in the UI, only the root cause.  If a user is not authenticated, they wouldn't be see this error anyways, they would get a 401.

@simonellistonball are you suggesting we move the logging call to the exception handler instead of the offending class?  In this case it's very likely an error would get logged twice.

Reply:
@cestella I initially tried to include the whole stack trace in the response but I couldn't find a good way to format it.  Multi-line statements just don't work well in JSON so I settled on the root cause because it's usually the most important piece of information anyways.  The whole stack trace is logged, just not returned in the JSON response.  Any ideas on how to do this better?

Reply:
I think the way you describe is the correct way

Reply:
@merrimanr I'm suggesting we log it somewhere, right now there is no apparent indication in the rest logs, so people end up hunting. When I hit the missing alter: nested template the other day, it took a PCAP of the transport protocol to find out what the actual root error was.

Reply:
@simonellistonball an error was not being logged before but is now in the PR.  Check out the ElasticsearchDao change.

Reply:
Hmm, in the situation that I was in last night while debugging, I was using this patch and the exception was not logged in metron-rest. I was hunting the same bug as Simon.  I ended up adjusting this to include the stack trace and that's how I eventually figured it out.  I'm not entirely certain WHY the error logging wasn't showing up.

Since this has happened to 2 of us now, could you validate that when you try to use the alerts-ui and use a sensor that does not have the `alert` nested template (any new sensor), you get an exception in the log?  It should be something like:
```
Caused by: org.apache.metron.indexing.dao.search.InvalidSearchException: Could not execute search
        at org.apache.metron.elasticsearch.dao.ElasticsearchDao.search(ElasticsearchDao.java:170)
        at org.apache.metron.elasticsearch.dao.ElasticsearchMetaAlertDao.search(ElasticsearchMetaAlertDao.java:190)
        at org.apache.metron.rest.service.impl.SearchServiceImpl.search(SearchServiceImpl.java:53)
        ... 88 more
Caused by: Failed to execute phase [query_fetch], all shards failed; shardFailures {[S6cJSn5nSmOpd7Lm2_OF9g][cowrie_index_2017.09.29.05][0]: RemoteTransportException[[node1][192.168.66.121:9300][indices:data/read/search[phase/query+fetch]]]; nested: SearchParseException[failed to parse search source [{\"from\":0,\"size\":25,\"query\":{\"constant_score\":{\"filter\":{\"bool\":{\"should\":[{\"query_string\":{\"query\":\"*\"}},{\"bool\":{\"must\":[{\"term\":{\"status\":\"active\"}},{\"nested\":{\"query\":{\"query_string\":{\"query\":\"*\"}},\"path\":\"alert\"}}]}}]}}}},\"_source\":{\"includes\":[],\"excludes\":[]},\"sort\":[{\"timestamp\":{\"order\":\"desc\"}}],\"track_scores\":true}]]; nested: QueryParsingException[[nested] failed to find nested object under path [alert]]; }
        at org.elasticsearch.action.search.AbstractSearchAsyncAction.onFirstPhaseResult(AbstractSearchAsyncAction.java:206)
        at org.elasticsearch.action.search.AbstractSearchAsyncAction$1.onFailure(AbstractSearchAsyncAction.java:152)
        at org.elasticsearch.action.ActionListenerResponseHandler.handleException(ActionListenerResponseHandler.java:46)
        at org.elasticsearch.transport.TransportService$DirectResponseChannel.processException(TransportService.java:855)
        at org.elasticsearch.transport.TransportService$DirectResponseChannel.sendResponse(TransportService.java:833)
        at org.elasticsearch.transport.TransportService$4.onFailure(TransportService.java:387)
        at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:39)
```
If you can get it to work, then I'm +1.  I absolutely couldn't get it to work with this PR, but I might've been doing something weird and I worry there's something weird with the logging level (which I think is at INFO) on metron-rest by default.

I still, personally, would rather see a stack trace in the error in the UI (since I'd prefer to *get* a stack trace from users for diagnosing issues and that's where they'll be looking).  I take @ottobackwards 's point about security, but I still think it's valid and valuable.  For the moment, I'd just append the multiline string and we can sort it out when debugging.  Mind you, this would be in addition to in the rest logs.

I guess the net-net of it is that I want it logged and validated at LEAST one place and I'd be ok if it were 2 (one to the user and one to the system logs).

Reply:
Also, yes, I realize that the stack trace shown is right in the path of your `LOG.error` call, but I absolutely did *not* see anything in the log.

Reply:
The log level is indeed INFO and it should not be.  That regression was introduced when we added ${HBASE_HOME}/conf to the REST classpath because that directory contains a log4j.properties file with logging set to INFO.  I'm planning on fixing that in a separate PR.  I'm not sure that's the cause here though because I'm able to produce an error in the logs with the same type of error (InvalidSearchException).

Are you sure you have the code in this PR deployed properly?  Your stack trace says the exception was thrown at line 170 but in this PR the exception is thrown at line 176.

Reply:
I'm ok with having it the UI, but it must be configurable or governed by the production profile.
You can have the customer switch and reproduce to get the screen shot ( although, when would you NOT want the whole log in practice? ).

Reply:
Yeah it’s could be a bad merge.  If you saw that in the log, then I’m ok with this.  My +1 stands.  Great job.